v2/api/keyvault/customizations/vault_extensions.go

```diff
@@ -0,0 +1,376 @@
+	if err != nil {
+		// Couldn't determine whether a soft-deleted keyvault exists, assume it doesn't
+
+		log.Error(err, "error checking for existence of soft-deleted KeyVault")
+		return nil
+	}
+
+	log.Info(
+		"KeyVault reconciliation requested PurgeThenCreate",
+		"KeyVault", kv.Name,
+		"Deleted KeyVault exists", exists)
+
+	if exists {
+		// if a soft-deleted KeyVault exists, we need to purge it before we can create a new one
+		req, err := ex.createPurgeRequest(ctx, kv, armClient, kubeClient, resolver, log)
+		if err != nil {
+			return errors.Wrapf(err, "error creating purge request for KeyVault %s", kv.Name)
+		}
+
+		_, err = armClient.Post(ctx, req)
+		if err != nil {
+			return errors.Wrapf(err, "error purging KeyVault %s", kv.Name)
+		}
+	}
+
+	return nil
+}
+
+// checkForExistenceOfDeletedKeyVault checks to see whether there's a soft deleted KeyVault with the same name.
+// This might be true if another party has deleted the KeyVault, even if we previously created it
+func (ex *VaultExtension) checkForExistenceOfDeletedKeyVault(
+	ctx context.Context,
+	kv *keyvault.Vault,
+	armClient *genericarmclient.GenericClient,
+	kubeClient kubeclient.Client,
+	resolver *resolver.Resolver,
+	log logr.Logger,
+) (bool, error) {
+	// Get the owner of the KeyVault, we need this resource group to determine the subscription
+	owner, ownerErr := ex.getOwner(ctx, kv, resolver, log)
+	if ownerErr != nil {
+		return false, errors.Wrapf(ownerErr, "unable to find owner of KeyVault %s", kv.Name)
+	}
+
+	// Get the subscription of the KeyVault
+	subscription, subscriptionErr := ex.getSubscription(owner)
+	if subscriptionErr != nil {
+		return false, errors.Wrapf(subscriptionErr, "unable to determine subscription for KeyVault %s", kv.Name)
+	}
+
+	// Get the location of the KeyVault
+	location, locationOk := ex.getLocation(kv, owner)
+	if !locationOk {
+		return false, errors.Errorf("unable to determine location of KeyVault %s", kv.Name)
+	}
+
+	// Create the ARM ID of the soft-deleted KeyVault, if any
+	armIdOfDeletedKeyVault, err := ex.createDeletedKeyVaultARMID(
+		subscription,
+		location,
+		kv.Name)
+	if err != nil {
+		return false, errors.Wrapf(err, "failed to create ARM ID for potential soft-deleted KeyVault %s", kv.Name)
+	}
+
+	// Get the API version of the KeyVault API to use
+	apiVersion, err := genruntime.GetAPIVersion(kv, kubeClient.Scheme())
+	if err != nil {
+		return false, errors.Wrapf(err, "unable to get api version for KeyVault")
+	}
+
+	// Check to see if the deleted keyvault exists
+	exists, _, err := armClient.HeadByID(ctx, armIdOfDeletedKeyVault, apiVersion)
+	if err != nil {
+		return false, errors.Wrapf(err, "error checking for existence of deleted KeyVault %s", kv.Name)
+	}
+
+	log.Info(
+		"Checking for existence of soft-deleted KeyVault",
+		"KeyVault", kv.Name,
+		"Subscription", subscription,
+		"Location", location,
+		"Exists", exists,
+	)
+
+	return exists, nil
+}
+
+func (ex *VaultExtension) createPurgeRequest(
```

Comment:
Rather than handcrafting the purge request here, I'd suggest using [BeginPurgeDeleted](https://github.com/Azure/azure-sdk-for-go/blob/main/sdk/resourcemanager/keyvault/armkeyvault/vaults_client.go#L588) method from azure-sdk-for-go